### PR TITLE
Bug/initial deposit revert

### PR DESCRIPTION
**Pull request by MilGard91**
## Context

BUG: When there is a deposit amount on the creation of the new staking pool, the msg.sender was the Cover Contract since the method was called from it. Revert happens since there is no balance on the contract itself.

Issue: https://github.com/NexusMutual/smart-contracts/issues/315

## Changes proposed in this pull request

Before transfer check the sender, if it is Cover Contract then do the transfer from the manager.

## Test plan

Test expanded with integration tests for creating staking pools with initial deposits. One for the private stake, and one for the public. Tests also include staking of the manager and one additional staker after pool creation.

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: test/integration/Cover/stakingPools.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
+const { ethers } = require('hardhat');
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:1]: name the file 
createStakingPool.js

that's what is being tested

File: test/integration/Cover/stakingPools.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+const { parseEther } = ethers.utils;
+
+function calculateTrancheId(lastBlock, period, gracePeriod) {
+  return Math.floor((lastBlock.timestamp + period + gracePeriod) / (91 * 24 * 3600));
+}
+
+const DEFAULT_POOL_FEE = '5';
+
+describe('Creating Staking Pools with funds', function () {
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:13]: describe('createStakingPool')

File: test/integration/Cover/stakingPools.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  it('should create a private staking pool', async function () {
+    const {
+      DEFAULT_PRODUCT_INITIALIZATION,
+      contracts: { cover, tk, tc },
+    } = this;
+    const [manager, staker] = this.accounts.members;
+    const lastBlock = await ethers.provider.getBlock('latest');
+    const trancheId = calculateTrancheId(lastBlock, period, gracePeriod);
+
+    const managerNXMBalance0 = await tk.balanceOf(manager.address);
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:37]: managerNXMBalanceBefore

File: test/integration/Cover/stakingPools.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    const tokenControllerBalance0 = await tk.balanceOf(tc.address);
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:38]: tokenControllerBalanceBefore

File: test/integration/Cover/stakingPools.js [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      true, // isPrivatePool,
+      DEFAULT_POOL_FEE, // initialPoolFee
+      DEFAULT_POOL_FEE, // maxPoolFee,
+      DEFAULT_PRODUCT_INITIALIZATION,
+      deposit, // depositAmount,
+      trancheId, // trancheId
+    );
+    const txResult = await tx.wait();
+
+    const managerNXMBalance1 = await tk.balanceOf(manager.address);
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:52]: managerNXMBalanceAfter

File: test/integration/Cover/stakingPools.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    const tokenControllerBalance1 = await tk.balanceOf(tc.address);
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:53]: tokenControllerBalanceAfter

File: test/integration/Cover/stakingPools.js [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    const stakingPoolCountAfter = await cover.stakingPoolCount();
+
+    assert.equal(managerNXMBalance1.toString(), managerNXMBalance0.sub(deposit).toString());
+    assert.equal(tokenControllerBalance1.toString(), tokenControllerBalance0.add(deposit).toString());
+    assert.equal(stakingPoolCountAfter.toNumber(), stakingPoolCountBefore.toNumber() + 1);
+
+    const stakingPoolAddress = txResult.events.pop().args.stakingPoolAddress;
+    const stakingPool = await ethers.getContractAt('IntegrationMockStakingPool', stakingPoolAddress);
+
+    const managerStakingPoolNFTBalance0 = await stakingPool.balanceOf(manager.address);
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:63]: managerStakingPoolNFTBalanceBefore

File: test/integration/Cover/stakingPools.js [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        tokenId: 0, // new position
+        destination: AddressZero,
+      },
+    ]);
+
+    await txRestake.wait();
+
+    const managerNXMBalance2 = await tk.balanceOf(manager.address);
+    const tokenControllerBalance2 = await tk.balanceOf(tc.address);
+    const managerStakingPoolNFTBalance1 = await stakingPool.balanceOf(manager.address);
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:79]: managerStakingPoolNFTBalanceAfter

File: test/integration/Cover/stakingPools.js [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    assert.equal(managerNXMBalance2.toString(), managerNXMBalance1.sub(deposit).toString());
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:80]: you saw the pattern use Before/After in naming

File: test/integration/Cover/stakingPools.js [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    const tx = await cover.connect(manager).createStakingPool(
+      manager.address,
+      false, // isPrivatePool,
+      DEFAULT_POOL_FEE, // initialPoolFee
+      DEFAULT_POOL_FEE, // maxPoolFee,
+      DEFAULT_PRODUCT_INITIALIZATION,
+      deposit, // depositAmount,
+      trancheId, // trancheId
+    );
+    const txResult = await tx.wait();
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:119]: afaik this is tx.wait is redundant

File: test/integration/Cover/stakingPools.js [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+    const managerNXMBalance1 = await tk.balanceOf(manager.address);
+    const tokenControllerBalance1 = await tk.balanceOf(tc.address);
+    const stakingPoolCountAfter = await cover.stakingPoolCount();
+
+    assert.equal(managerNXMBalance1.toString(), managerNXMBalance0.sub(deposit).toString());
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:56]: don't do this toString() conversion

```    expect(ethBalanceAfter).to.be.equal(ethBalanceBefore.add(segment.amount));``` 

use this syntax, should work

Comment (danoctavian) [reply test/integration/Cover/stakingPools.js:56]: everywhere

File: test/integration/Cover/stakingPools.js [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+          trancheId,
+          tokenId: 0, // new position
+          destination: AddressZero,
+        },
+      ]),
+      'StakingPool: The pool is private',
+    );
+  });
+
+  it('should create a public staking pool', async function () {
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:97]: should create a public staking pool with an initial deposit

File: test/integration/Cover/stakingPools.js [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    const { tk } = this.contracts;
+
+    const members = this.accounts.members.slice(0, 5);
+    const amount = parseEther('10000');
+    for (const member of members) {
+      await tk.connect(this.accounts.defaultSender).transfer(member.address, amount);
+    }
+  });
+
+  it('should create a private staking pool', async function () {
```
Comment (danoctavian) [inline test/integration/Cover/stakingPools.js:28]: should create a private staking pool with an initial deposit

File: contracts/modules/token/NXMToken.sol [hunk 14]
```diff
@@ -218,7 +218,7 @@ contract NXMToken is OZIERC20 {
   * @param value The amount to be transferred.
   */
   function operatorTransfer(address from, uint256 value) public onlyOperator returns (bool) {
-    require(value <= _balances[from]);
+    require(value <= _balances[from], "Insufficient funds to allow transfer");
```
Comment (shark0der) [inline contracts/modules/token/NXMToken.sol:221]: The NXMToken contract is immutable and we're "stuck" with it as it is, hence we don't change the code nor compiler version for it. The deps are also copied/frozen.

File: test/integration/Cover/createStakingPool.js [hunk 15]
```diff
@@ -0,0 +1,169 @@
+const { ethers } = require('hardhat');
+const { expect } = require('chai');
+const { expectRevert } = require('@openzeppelin/test-helpers');
```
Comment (shark0der) [inline test/integration/Cover/createStakingPool.js:3]: We're switching from oz test helpers to waffle/chai. Remove this import

File: test/integration/Cover/createStakingPool.js [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+        {
+          amount: deposit,
+          trancheId,
+          tokenId: 0, // new position
+          destination: AddressZero,
+        },
+      ]),
+      'StakingPool: The pool is private',
+    );
+  });
```
Comment (shark0der) [inline test/integration/Cover/createStakingPool.js:93]:
```
await expect(stakingPool.connect(staker).depositTo([
        {
          amount: deposit,
          trancheId,
          tokenId: 0, // new position
          destination: AddressZero,
        },
      ])).to.be.revertedWith('StakingPool: The pool is private');
```

File: test/integration/Cover/createStakingPool.js [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    );
+    const managerNXMBalanceAfterCreation = await tk.balanceOf(manager.address);
+    const tokenControllerBalanceAfterCreation = await tk.balanceOf(tc.address);
+    const stakingPoolCountAfter = await cover.stakingPoolCount();
+
+    expect(managerNXMBalanceAfterCreation).to.be.equal(managerNXMBalanceBefore.sub(deposit));
+    expect(tokenControllerBalanceAfterCreation).to.be.equal(tokenControllerBalanceBefore.add(deposit));
+    expect(stakingPoolCountAfter).to.be.equal(stakingPoolCountBefore.add(1));
+
+    const stakingPoolAddress = await cover.stakingPool(stakingPoolCountAfter - 1);
```
Comment (shark0der) [inline test/integration/Cover/createStakingPool.js:59]: I'm expecting `stakingPoolCountAfter` to be a BigNumber, use `.sub` instead of  `-`.

**Review by shark0der — CHANGES_REQUESTED** [review]
Left my comments on the code :)

**Review by roxdanila — APPROVED** [review]

**Review by shark0der — APPROVED** [review]

**Pull request merged**